### PR TITLE
[Bugfix][MoE] Preserve unquantized weight storage on ROCm

### PR DESCRIPTION
**Pull request by aaab8b**
## Summary
- Preserve unquantized MoE parameter storage during ROCm post-load padding by copying the padded view into existing storage instead of rebinding `.data`.
- Return directly after AITER unquantized MoE `shuffle_weights()` so the AITER runtime layout is not forced through an extra `.contiguous()` conversion.
- Add a ROCm MoE weight replay test covering both AITER and TRITON backends.

## Motivation
In RL-style serving and training loops, external weight updates can trigger `process_weights_after_loading()` more than once after model initialization. Reassigning `param.data` in that path can replace the underlying parameter storage. That is unsafe once CUDA graphs or runtime kernels have captured parameter storage addresses, and can leave later rollouts reading stale or invalid storage.

We observed this as a training-time rollout corruption issue after weight updates: generated responses became invalid, and reward/score collapsed to 0. Preserving the existing parameter storage with `.copy_()` keeps captured addresses stable while still applying the padded layout.

For AITER unquantized MoE, `shuffle_weights()` already returns the runtime layout consumed by AITER. In local Qwen3.5-35B GRPO measurements, avoiding the final contiguous conversion for AITER was slightly faster after normalizing rollout generation time by response length.

## Replay test
The new `tests/rocm/test_moe_weight_replay.py` test constructs a Qwen3.5 MoE LLM with sleep mode and CUDA graphs, generates a deterministic greedy response, runs level-2 sleep/wake, then replays checkpoint-format weights through a framework-agnostic load-then-post-load helper:

1. stream checkpoint-format weights into `model.load_weights(...)` in buckets;
2. call `process_weights_after_loading(...)` once after all buckets are loaded;
3. verify the ROCm MoE padding branch was actually exercised;
4. verify MoE `w13_weight` / `w2_weight` parameter storage addresses do not change;
5. verify greedy token IDs are deterministic after replay.

This covers both AITER and TRITON MoE backends as first-class correctness cases.

## Local validation
Static/lightweight checks in the PR checkout:
- `python3 -m py_compile tests/rocm/test_moe_weight_replay.py vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py vllm/model_executor/layers/fused_moe/oracle/unquantized.py`
- `python3 -m ruff check tests/rocm/test_moe_weight_replay.py vllm/model_executor/layers/fused_moe/unquantized_fused_moe_method.py vllm/model_executor/layers/fused_moe/oracle/unquantized.py`
- `python3 -m pytest tests/kernels/moe/test_unquantized_backend_selection.py tests/kernels/moe/test_moe_weight_loading_padded.py`
  - 38 passed, 2 skipped

Heavy replay validation with Qwen3.5-35B-A3B and tensor parallel size 2:
- pre-fix AITER: failed as expected; the test entered the ROCm padding branch and detected MoE parameter `data_ptr` changes.
- pre-fix TRITON: failed as expected; the test entered the ROCm padding branch and detected MoE parameter `data_ptr` changes.
- fixed AITER: passed.
- fixed TRITON: passed.

## Benchmark note
Local 10-step Qwen3.5-35B GRPO runs kept score/reward normal with zero aborted responses. Normalized rollout timing for AITER was slightly better without the final contiguous conversion: about 22.95 ms/token vs 23.11 ms/token.

### CODE REVIEW

**Comment by AndreasKaratzas** [comment]
cc @Rohan138

File: tests/rocm/test_moe_weight_replay.py [hunk 1]
```diff
@@ -0,0 +1,326 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (tjtanaa) [inline tests/rocm/test_moe_weight_replay.py:1]: This test doesn't seem to run

# pytest -svvv tests/rocm/test_moe_weight_replay.py | tee moe.log
============================= test session starts ==============================
platform linux -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /app/reviewpr06/fix-aiter-moe-weight-storage
configfile: pyproject.toml
plugins: asyncio-1.4.0, anyio-4.14.1
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collecting ... collected 2 items

tests/rocm/test_moe_weight_replay.py::test_moe_weight_replay_after_sleep_wake_is_deterministic[triton] SKIPPED
tests/rocm/test_moe_weight_replay.py::test_moe_weight_replay_after_sleep_wake_is_deterministic[aiter] SKIPPED

**Comment by tjtanaa** [comment]
@aaab8b  And for further review, Please fix the pre-commit first

File: tests/rocm/test_moe_weight_replay.py [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+aiter_available = importlib.util.find_spec("aiter") is not None
+
+pytestmark = pytest.mark.skipif(
+    not (current_platform.is_rocm() and aiter_available),
+    reason="MoE replay test requires ROCm with AITER installed",
+)
+
+
+def _visible_gpu_count() -> int:
+    visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES")
```
Comment (tjtanaa) [inline tests/rocm/test_moe_weight_replay.py]: We should check for `HIP_VISIBLE_DEVICES`, instead. We will be deprecating `CUDA_VISIBLE_DEVICES` alias on v0.26.0

File: tests/rocm/test_moe_weight_replay.py (hunk 1 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py:1]: maybe you can run it with -rs, to show what reason causes the skip.

File: tests/rocm/test_moe_weight_replay.py [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+        gc.collect()
+        torch.cuda.empty_cache()
+
+
+@pytest.mark.parametrize("backend", ["triton", "aiter"])
+def test_moe_weight_replay_after_sleep_wake_is_deterministic(
+    monkeypatch: pytest.MonkeyPatch,
+    backend: str,
+) -> None:
+    tp_size = int(os.environ.get(TP_ENV, "8"))
```
Comment (tjtanaa) [inline tests/rocm/test_moe_weight_replay.py]: @aaab8b This TP size is too large. For this test, we either go for TP1 or TP2.

File: tests/rocm/test_moe_weight_replay.py (hunk 1 above)
Comment (tjtanaa) [reply tests/rocm/test_moe_weight_replay.py:1]: We should make sure that this is picked up by `AMD CI` and is running rather than being skipped.
If the test is not relevant to any of the Test Group in `test-amd.yaml`, please create a new Test Group.

CC @AndreasKaratzas

Comment (tjtanaa) [reply tests/rocm/test_moe_weight_replay.py:1]: @aaab8b 

```
tests/rocm/test_moe_weight_replay.py ss                                  [100%]

=============================== warnings summary ===============================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

tests/rocm/test_moe_weight_replay.py: 14 warnings
  /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
SKIPPED [2] tests/rocm/test_moe_weight_replay.py:266: MoE C extension with topk_softmax is required
```

File: tests/rocm/test_moe_weight_replay.py (hunk 3 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py]: done. Changing default to TP2. And I've tested the script. the script works as well.

File: tests/rocm/test_moe_weight_replay.py (hunk 2 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py]: Done.

**Comment by aaab8b** [comment]
> @aaab8b And for further review, Please fix the pre-commit first

Done.fixed the pre-commit.

**Comment by AndreasKaratzas** [comment]
I removed ready label until DCO is fixed.

File: tests/rocm/test_moe_weight_replay.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+        monkeypatch.setenv("VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION", "0")
+    elif backend == "triton":
+        monkeypatch.delenv("VLLM_ROCM_USE_AITER", raising=False)
+        monkeypatch.setenv("VLLM_ROCM_USE_AITER_MOE", "0")
+    else:
+        raise AssertionError(f"unknown backend {backend}")
+
+    from vllm._aiter_ops import rocm_aiter_ops
+
+    rocm_aiter_ops.refresh_env_variables()
```
Comment (AndreasKaratzas) [inline tests/rocm/test_moe_weight_replay.py:77]: What is going on here? First of all use V1 does not exist. I mean I see some flags here that are equal to defaults anyway (eg VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION).

File: tests/rocm/test_moe_weight_replay.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            llm.shutdown()
+        elif hasattr(llm.llm_engine, "shutdown"):
+            llm.llm_engine.shutdown()
+        elif hasattr(llm.llm_engine, "engine_core"):
+            llm.llm_engine.engine_core.shutdown()
+    finally:
+        del llm
+        gc.collect()
+        torch.accelerator.empty_cache()
+
```
Comment (AndreasKaratzas) [inline tests/rocm/test_moe_weight_replay.py]: What is this? Why do we need all of that?

**Review by AndreasKaratzas — COMMENTED** [review]
I think this test is badly written. I don't even know where to start here.

File: tests/rocm/test_moe_weight_replay.py (hunk 1 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py:1]: > [@aaab8b](https://github.com/aaab8b)
> 
> ```
> tests/rocm/test_moe_weight_replay.py ss                                  [100%]
> 
> =============================== warnings summary ===============================
> <frozen importlib._bootstrap>:488
>   <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
> 
> <frozen importlib._bootstrap>:488
>   <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute
> 
> tests/rocm/test_moe_weight_replay.py: 14 warnings
>   /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
>     warnings.warn(
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> =========================== short test summary info ============================
> SKIPPED [2] tests/rocm/test_moe_weight_replay.py:266: MoE C extension with topk_softmax is required
> ```

You need aiter to use aiter backend test.

File: tests/rocm/test_moe_weight_replay.py (hunk 4 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py:77]: I've clearly declared why I'm doing this in the PR main comment. You can read it first. I'm just making sure it uses V1. For those envs like VLLM_ROCM_USE_AITER and VLLM_ROCM_USE_AITER, I used it for testing on two backend for MoE both on triton and aiter backend. Other useless environment settings have been moved as you suggested.

File: tests/rocm/test_moe_weight_replay.py (hunk 5 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py]: Just shut down the llm instance and clear the resources on ROCm.

File: tests/rocm/test_moe_weight_replay.py (hunk 1 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py:1]: > [@aaab8b](https://github.com/aaab8b)
> 
> ```
> tests/rocm/test_moe_weight_replay.py ss                                  [100%]
> 
> =============================== warnings summary ===============================
> <frozen importlib._bootstrap>:488
>   <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
> 
> <frozen importlib._bootstrap>:488
>   <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute
> 
> tests/rocm/test_moe_weight_replay.py: 14 warnings
>   /usr/local/lib/python3.12/dist-packages/torch/jit/_script.py:365: DeprecationWarning: `torch.jit.script_method` is deprecated. Please switch to `torch.compile` or `torch.export`.
>     warnings.warn(
> 
> -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
> =========================== short test summary info ============================
> SKIPPED [2] tests/rocm/test_moe_weight_replay.py:266: MoE C extension with topk_softmax is required
> ```

@tjtanaa I've discovered that it is needed to import vllm._moe_C_stable_libtorch first so that the script can have MoE C extension. I've tested this on rocm-ci image rocm/vllm-ci:f7efab58ec27. This script works fine now.

Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py:1]: > We should make sure that this is picked up by `AMD CI` and is running rather than being skipped. If the test is not relevant to any of the Test Group in `test-amd.yaml`, please create a new Test Group.
> 
> CC [@AndreasKaratzas](https://github.com/AndreasKaratzas)

I've added it into test-amd.yaml with aiter ops part. This is still a MoE aiter implementation associated problem. @AndreasKaratzas

File: tests/rocm/test_moe_weight_replay.py (hunk 4 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py:77]: > What is going on here? First of all use V1 does not exist. I mean I see some flags here that are equal to defaults anyway (eg VLLM_ROCM_USE_AITER_UNIFIED_ATTENTION).

@AndreasKaratzas I have rewritten the test script. Now it only needs to set VLLM_ROCM_USE_AITER and VLLM_ROCM_USE_AITER_MOE for envs.

File: tests/rocm/test_moe_weight_replay.py (hunk 5 above)
Comment (aaab8b) [reply tests/rocm/test_moe_weight_replay.py]: > What is this? Why do we need all of that?

@AndreasKaratzas all shutdown llm instance codes are removed. In fact, in this new test, we don't even need to initialize a llm instance now.

File: vllm/model_executor/layers/fused_moe/oracle/unquantized.py [hunk 6]
```diff
@@ -303,13 +304,6 @@ def convert_to_unquantized_kernel_format(
             is_gated_act_gemm=is_act_and_mul,
         )
 
-    if (
-        unquantized_backend == UnquantizedMoeBackend.TRITON
-        and current_platform.is_rocm()
-        and envs.VLLM_ROCM_MOE_PADDING
-    ):
-        # Skip .contiguous(): it would undo the ROCm MoE weight padding.
-        return w13_weight, w2_weight
```
Comment (aaab8b) [inline vllm/model_executor/layers/fused_moe/oracle/unquantized.py:370]: We've described the speed result from end to end. Aiter backend needs certain padding to speed-up a little bit in rocm.

**Review by tjtanaa — APPROVED** [review]

**Comment by tjtanaa** [comment]
@aaab8b can you help to resolve the merge conflict ?

**Comment by aaab8b** [comment]
> @aaab8b can you help to resolve the merge conflict ?

@tjtanaa Sure. Let me check this.

**Comment by tjtanaa** [comment]
@aaab8b could you help to resolve this merge conflict ?

**Comment by tjtanaa** [comment]
I will try to get this merge this week.

**Comment by aaab8b** [comment]
> I will try to get this merge this week.

@tjtanaa thanks！ I've done solving this merge conflict.
